components/brave_rewards/browser/rewards_service_impl.h

```diff
@@ -605,6 +620,9 @@ class RewardsServiceImpl : public RewardsService,
   mojo::AssociatedRemote<bat_ledger::mojom::BatLedger> bat_ledger_;
   mojo::Remote<bat_ledger::mojom::BatLedgerService> bat_ledger_service_;
   const scoped_refptr<base::SequencedTaskRunner> file_task_runner_;
+#if BUILDFLAG(IS_ANDROID)
+  const scoped_refptr<base::SequencedTaskRunner> json_sanitizer_task_runner_;
```

Comment:
Did you consider having all platforms run sanitization on a different task runner, for simplicity? Is that possible, and if so, what would be the downsides?

Reply:
@zenparsing , I was considering that but in the case of Desktop it would be redundant.
The thing is that on Desktop sanitizing already happens in other thread or maybe even sandbox process [[1]](https://source.chromium.org/chromium/chromium/src/+/main:services/data_decoder/public/cpp/json_sanitizer_non_android.cc;l=18).
And on Android sanitizing happens in the same thread as caller [[2]](https://source.chromium.org/chromium/chromium/src/+/main:services/data_decoder/public/cpp/json_sanitizer_android.cc;l=52) , in Java code. 
So making the same for the desktop would cost unnecessary class member and unnecessary task post, and I decided to separate Android implementation.

Reply:
Yeah, I agree - not a huge fan of conditional compilations. Can we measure the delay (if any) this would add @AlexeyBarabash?

Reply:
sure, let me run the check the build with checks.

Reply:
@szilardszaloki  @zenparsing  the difference surprises me, desktop way is ~40x faster:

Desktop way: ~0.05 sec
Android way: ~2 sec

The draft code which I used: https://github.com/brave/brave-core/commit/a3e8f87393bd6e903aec6d79ca7b9fb4a98b9fc0 .



<details>
  <summary>Details from my logs:</summary>
 
  ```
[2922601:2922601:0225/001041.947591:ERROR:rewards_service_impl.cc(1176)] [Brave] JsonSanitizeCallback DESKTOP WAY: 68457 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
[2922648:2922648:0225/001041.947886:ERROR:get_parameters.cc(133)] [Brave] ProcessResponse 000
[2922648:2922648:0225/001041.947904:ERROR:get_parameters.cc(23)] [Brave] ParseBody 000
[2922601:2922601:0225/001042.140115:ERROR:rewards_service_impl.cc(1176)] [Brave] JsonSanitizeCallback DESKTOP WAY: 57965 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2922601:2922601:0225/001042.335399:ERROR:rewards_service_impl.cc(1176)] [Brave] JsonSanitizeCallback DESKTOP WAY: 49668 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux


[2933268:2933268:0225/002333.666193:ERROR:rewards_service_impl.cc(1180)] [Brave] JsonSanitizeCallback DESKTOP WAY: 57163 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
[2933316:2933316:0225/002333.666482:ERROR:get_parameters.cc(133)] [Brave] ProcessResponse 000
[2933316:2933316:0225/002333.666498:ERROR:get_parameters.cc(23)] [Brave] ParseBody 000
[2933268:2933268:0225/002334.215691:ERROR:rewards_service_impl.cc(1180)] [Brave] JsonSanitizeCallback DESKTOP WAY: 37018 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2933268:2933268:0225/002334.259963:ERROR:rewards_service_impl.cc(1180)] [Brave] JsonSanitizeCallback DESKTOP WAY: 63543 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux



[2927529:2927529:0225/001649.974140:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2054619 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2927529:2927529:0225/001649.976225:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2036463 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2927529:2927529:0225/001649.980277:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2516733 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters


[2929901:2929901:0225/001937.085739:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2377420 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2929901:2929901:0225/001937.089820:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2252655 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[2929901:2929901:0225/001937.104630:ERROR:rewards_service_impl.cc(1170)] [Brave] operator() ANDROID WAY: 2220402 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
  
  ```
</details>

Reply:
Oh, that's quite bad. Actually, it's such a bad performance hit that I'd even question whether we need this on Android.

Reply:
I think it's expected that it will be slower. The thing is that the way like it is right now, makes it visible that UI won't redraw for user and it gives `App Not Responding` on mobile. With the way @AlexeyBarabash did the app will be reliable and responsive, just a bit longer time for rewards to draw something in UI and that's it. Not sure why it's bad @szilardszaloki. How many times do we do such request if we open a rewards panel on a particular web site, I think it's better to show some info with a delay when app is in the responsive state comparing to show faster with an App Not Responding Android OS system message.

Reply:
@SergeyZhukovsky Agreed that the `App Not Responding` should go. 🙂 I'm just a bit concerned that a 2-sec delay in the sanitizer is just too much in general, given that all our network responses go through it.

Reply:
cc: @spylogsster to see if he has anything to say about it. Also curious about what @zenparsing thinks.

Reply:
Here are the same tested on Android, Pixel 6:

```
02-27 12:31:40.453: E/chromium(10656): [ERROR:rewards_service_impl.cc(1180)] [Brave] JsonSanitizeCallback DESKTOP WAY: 1553 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=beb06bd1-a078-5922-a215-9bda690865f1&platform=android
...

02-27 12:46:53.885: E/chromium(12799): [ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY: 658 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
```

The non-UI thread is slower, but not so dramatically 0.001553 sec vs 0.000658 sec .

2 sec on Desktop seems as the value not supposed to be. At least because my desktop device is much faster than the phone.

I am redoing measuring with existing [file_task_runner_](https://github.com/brave/brave-core/pull/17354/files#diff-df0abf3b2c26fef46c7ac0d8b8d6e22d344e8b69576dcf33faed8c595b1cdc0dR622) instead of introducing new `json_sanitizer_task_runner_`.

Reply:
With re-using `file_task_runner_`:
Android device: ~0.002..0.0008 sec
Desktop device: ~0.2..0.06 sec

-  it is not 2 sec as before, it looks like I need to re-measure desktop with `json_sanitizer_task_runner_`

<details>
  <summary>Details from my logs:</summary>

```
Reuse PIXEL6

02-27 14:07:37.433: E/chromium(22219): [ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 2065 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
02-27 14:08:07.259: E/chromium(22219): [ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 800 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=beb06bd1-a078-5922-a215-9bda690865f1&platform=android


====

Reuse DESKTOP

[99230:99230:0227/141941.932010:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 224150 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
[99230:99230:0227/141941.932123:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 249452 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[99230:99230:0227/141941.932199:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 152065 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux


[99230:99230:0227/142011.187522:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 67188 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[99230:99230:0227/142012.019866:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(file_task_runner_): 62835 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux


  ```
</details>

Reply:
I had run that again, desktop, with json_sanitizer_task_runner_, before the test I rebooted the PC and closed all except terminal and VS code to avoid negative influence as possible.

Only the first call takes 2.5 sec, all other calls take 0.05 sec.  Which (0.05 sec) is similar to pre-PR measurement.

So it looks my very first measurement was wrong and we can use the similar way for desktop and Android by sacrificing only 1st call.

```
[17916:17916:0227/150744.056809:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(json_sanitizer_task_runner_): 2469718 microseconds URL=https://api.rewards.bravesoftware.com/v1/parameters
[17964:17964:0227/150744.057112:ERROR:get_parameters.cc(133)] [Brave] ProcessResponse 000
[17964:17964:0227/150744.057133:ERROR:get_parameters.cc(23)] [Brave] ParseBody 000
[17916:17916:0227/150745.070996:ERROR:object_proxy.cc(623)] Failed to call method: org.freedesktop.ScreenSaver.GetActive: object_path= /org/freedesktop/ScreenSaver: org.freedesktop.DBus.Error.NotSupported: This method is not implemented
[17916:17916:0227/150746.860301:WARNING:CONSOLE(161)] "crbug/1173575, non-JS module files deprecated.", source: chrome://resources/js/load_time_data_deprecated.js (161)
[17916:17916:0227/150747.713872:WARNING:CONSOLE(1)] "Could not find style data in module named null", source: chrome://resources/polymer/v3_0/polymer/polymer_bundled.min.js (1)
[17916:17916:0227/150747.957499:WARNING:CONSOLE(1)] "Could not find style data in module named null", source: chrome://resources/polymer/v3_0/polymer/polymer_bundled.min.js (1)
[17916:17916:0227/150747.984808:WARNING:CONSOLE(1)] "Could not find style data in module named null", source: chrome://resources/polymer/v3_0/polymer/polymer_bundled.min.js (1)
[17916:17916:0227/150748.037528:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(json_sanitizer_task_runner_): 40462 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[17916:17916:0227/150748.065810:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(json_sanitizer_task_runner_): 48495 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux
[17916:17916:0227/150748.089410:ERROR:CONSOLE(2145)] "[Brave Settings Overrides] Appearance Page cannot find use default theme button template", source: chrome://settings/settings.js (2145)
[17916:17916:0227/150811.045648:ERROR:rewards_service_impl.cc(1174)] [Brave] operator() ANDROID WAY(json_sanitizer_task_runner_): 43114 microseconds URL=https://grant.rewards.bravesoftware.com/v1/promotions?migrate=true&paymentId=b7d1cd25-794d-5a14-a6d9-fb7491b69687&platform=linux

```

Reply:
Sounds good - thanks for doing all the above measurements! 🙂

Reply:
I have made code similar for different platforms: 
https://github.com/brave/brave-core/pull/17354/commits/b840262638503b72fff61ecc3e827be13575aa27